src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenNullCoalescingAssignmentTests.cs

```diff
@@ -1580,7 +1941,376 @@ static void M(dynamic d1, dynamic d2)
     }
 }";
 
-            var verifier = CompileAndVerify(source, new[] { CSharpRef }, expectedOutput: "");
+            CompileAndVerify(source, new[] { CSharpRef }, expectedOutput: "");
+        }
+
+        [Fact]
+        public void AwaitLHS()
+        {
+            var source = @"
+using System.Threading.Tasks;
+class C
+{
+    static async Task M(Task<Task> t1, Task t2)
+    {
+        await t1 ??= t2;
+    }
+}
+";
+
+            CreateCompilation(source).VerifyDiagnostics(new DiagnosticDescription[] {
+                // (7,9): error CS0131: The left-hand side of an assignment must be a variable, property or indexer
+                //         await t1 ??= t2;
+                Diagnostic(ErrorCode.ERR_AssgLvalueExpected, "await t1").WithLocation(7, 9)
+            });
+        }
+
+        [Fact]
+        public void OperatorEqualsIgnored()
+        {
+            var source = @"
+using System;
+class C
+{
+    public static bool operator==(C c1, C c2)
+    {
+        Console.WriteLine(""In operator=="");
+        return false;
+    }
+    public static bool operator!=(C c1, C c2)
+    {
+        Console.WriteLine(""In operator!="");
+        return false;
+    }
+    public override bool Equals(object o)
+    {
+        Console.WriteLine(""In C.Equals"");
+        return false;
+    }
+    public override int GetHashCode() => 1;
+
+    public static void Main()
+    {
+        C c1 = null, c2 = new C();
+        c1 ??= c2;
+        Console.WriteLine(c1.GetHashCode());
+    }
+}";
+            var verifier = CompileAndVerify(source, expectedOutput: "1");
+            verifier.VerifyIL("C.Main()", expectedIL: @"
+{
+  // Code size       25 (0x19)
+  .maxstack  1
+  .locals init (C V_0, //c1
+                C V_1) //c2
+  IL_0000:  ldnull
+  IL_0001:  stloc.0
+  IL_0002:  newobj     ""C..ctor()""
+  IL_0007:  stloc.1
+  IL_0008:  ldloc.0
+  IL_0009:  brtrue.s   IL_000d
+  IL_000b:  ldloc.1
+  IL_000c:  stloc.0
+  IL_000d:  ldloc.0
+  IL_000e:  callvirt   ""int object.GetHashCode()""
+  IL_0013:  call       ""void System.Console.WriteLine(int)""
+  IL_0018:  ret
+}
+");
+        }
+
+        [Fact]
+        public void Associativity()
+        {
+            var source = @"
+using System;
+class C
+{
+    public static void Main()
+    {
+        int? a1 = null, b1 = null, c1 = 1;
+        Console.WriteLine(a1 ??= b1 ??= c1);
```

Comment:
Would the result be any different if this were executed as `(a1 ??= b1) ??= c1`? Consider verifying the IL so it's more obvious.

Reply:
`a ??= b` is not an lvalue, so this scenario cannot exist.

---
In reply to: [216109782](https://github.com/dotnet/roslyn/pull/29691#discussion_r216109782) [](ancestors = 216109782)

Reply:
In that case, consider verifying the intermediate values as well:
```
Console.WriteLine(a1 ??= b1 ??= c1);
Console.WriteLine(b1);
```

---
In reply to: [216397478](https://github.com/dotnet/roslyn/pull/29691#discussion_r216397478) [](ancestors = 216397478,216109782)

Reply:
I'll address this in the followup

---
In reply to: [216759775](https://github.com/dotnet/roslyn/pull/29691#discussion_r216759775) [](ancestors = 216759775,216397478,216109782)